api/bucket.go

```diff
@@ -1,19 +1,205 @@
 package api
 
 import (
-	"github.com/gin-gonic/gin"
-	"github.com/konveyor/tackle2-hub/model"
+	"archive/tar"
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"fmt"
 	"io"
+	"io/ioutil"
 	"net/http"
 	"os"
+	"path"
 	pathlib "path"
+	"path/filepath"
+	"strings"
+
+	"github.com/gin-gonic/gin"
+	"github.com/konveyor/tackle2-hub/model"
 )
 
 //
 // BucketHandler provides bucket management.
 type BucketHandler struct {
 }
 
+// TODO: Validate Wildcard param to not allow access dirs outside of the bucket
+func (h *BucketHandler) serveBucketGet(ctx *gin.Context, owner *model.BucketOwner) {
+	if ctx.Request.Header.Get(Accept) == TarGzMimetype {
+		h.getDirArchive(ctx, path.Join(owner.Bucket, ctx.Param(Wildcard)))
+	} else {
+		h.content(ctx, owner)
+	}
+}
+
+func (h *BucketHandler) serveBucketUpload(ctx *gin.Context, owner *model.BucketOwner) {
+	if ctx.Request.Method == "PUT" {
+		h.uploadDirArchive(ctx, path.Join(owner.Bucket, ctx.Param(Wildcard)))
+	} else {
+		h.upload(ctx, owner)
+	}
+}
+
+func (h *BucketHandler) uploadDirArchive(ctx *gin.Context, dir string) {
+	// Prepare to uncompress the uploaded data, report 4xx errors
+	file, err := ctx.FormFile("file")
+	if err != nil {
+		ctx.Status(http.StatusBadRequest)
+		return
+	}
+	fileReader, err := file.Open()
+	if err != nil {
+		ctx.Status(http.StatusBadRequest)
+		return
+	}
+	defer fileReader.Close()
+
+	ungzReader, err := gzip.NewReader(fileReader)
+	if err != nil {
+		ctx.Status(http.StatusBadRequest)
+		return
+	}
+	defer ungzReader.Close()
+
+	// Report 5xx errors in upcoming steps
+	defer func() {
+		if err != nil {
+			ctx.JSON(http.StatusInternalServerError, gin.H{
+				"error": err,
+			})
+			return
+		}
+	}()
+
+	// Prepare destionation directory
+	bucketContent, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return
+	}
+	for _, bucketEntry := range bucketContent {
+		err = os.RemoveAll(path.Join(dir, bucketEntry.Name()))
+		if err != nil {
+			return
+		}
+	}
+
+	// Extract the tar archive
+	untarReader := tar.NewReader(ungzReader)
+	for {
+		hdr, err := untarReader.Next()
+		if err != nil {
+			if err == io.EOF {
+				break
+			} else {
+				return
+			}
+		}
+
+		switch hdr.Typeflag {
+		case tar.TypeDir:
+			if err := os.Mkdir(path.Join(dir, hdr.Name), hdr.FileInfo().Mode().Perm()); err != nil {
+				return
+			}
+		case tar.TypeReg:
+			var file *os.File
+			if file, err = os.Create(path.Join(dir, hdr.Name)); err != nil {
+				return
+			}
+			if _, err = io.Copy(file, untarReader); err != nil {
+				return
+			}
+			file.Close()
+		default:
+			// types that are not files or dirs are skipped
+		}
+	}
+
+	ctx.Status(http.StatusAccepted)
+}
+
```

Comment:
The getDirArchive method takes a gin ctx as a destination where it sends the streamed archive data and a string specifying a directory that should be archived.